src/Lean/Meta/KAbstract.lean

```diff
@@ -41,15 +43,22 @@ def kabstract (e : Expr) (p : Expr) (occs : Occurrences := .all) : MetaM Expr :=
         visitChildren ()
       else if e.toHeadIndex != pHeadIdx || e.headNumArgs != pNumArgs then
         visitChildren ()
-      else if (← isDefEq e p) then
-        let i ← get
-        set (i+1)
-        if occs.contains i then
-          return mkBVar offset
+      else
+        -- We save the metavariable context here,
+        -- so that it can be rolled back unless `occs.contains i`.
+        let mctx ← getMCtx
```

Comment:
I noticed that in the definition of `checkPointDefEq`, it uses the `MonadBacktrack SavedState MetaM` instance for restoring the state if `isDefEq` fails or returns false. Is there any reason that you instead use the `MonadMCtx MetaM` instance?